scipy/integrate/_ivp/ivp.py

```diff
@@ -652,7 +652,15 @@ def fun(t, x, fun=fun):
 
     status = None
     while status is None:
-        message = solver.step()
+        try:
+            message = solver.step()
+        except Exception as e:
```

Comment:
It looks like @nmayorov has touched a lot of the flanking code, so pinging them makes sense I think.

Is there any precedent for this type of approach in SciPy? Catching the generic exception class is of course a bit ugly, and we'd presumably want a test for this functionality if it is desirable (is it? I don't know enough to comment).

Reply:
There is plenty of precedent in `optimize` for returning gracefully with a failure status/message. Usually this is done by checking for some failure condition rather than by intercepting an `Exception`, but I can think of an... exception:

gh-14858 showed that `bracket` could fail silently and cause scalar minimization functions to return nonsense. The desired behavior was for `minimize_scalar` to return gracefully and report the failure status, and the cleanest way to retrofit that was to make`bracket` raise an error and for `minimize_scalar` to intercept it. (Otherwise, a lot of return value routing would have been required.) 

It sounds like that's similar to what this PR wants to do, so it seems OK to me as long as `success` is `False` and the `status` code is something appropriate (`-1` or something new).

Reply:
Thanks for the reviews!

I agree that catching the generic exception class is ugly and can cause problems. The issue that pushed me to this, though, is that there could be any number of types of exceptions that can be raised, since a user-provided function `func` is evaluated at each time step. There's just no way to know beforehand what sort of exceptions may be raised by calls to `func`. So it seemed to me to be easiest to just catch everything.

The only alternative that I can think of would be to allow the user to specify which sort of exception (or maybe a list of exception types?) that they want to catch. Then only catch those. But that seems like more work for the user.

The main issue that arises when catching all exceptions with `except Exception as e:` is that it can make code silently fail. Here however, the failure is not silent, since the solver will stop and the exception string is incorporated into the message that the solver returns. Plus, it's easy enough for the user to rerun with `catch_exceptions=False`, in order to see the original exception in its proper place in the stack.

Also note that I did not use `status = -1`, because when I tried that my custom message was overwritten later in the function: `message = MESSAGES.get(status, message)`. Hopefully `status = -2` is sufficiently clear and different to work well here.